service/service.go

```diff
@@ -279,168 +328,145 @@ func (s *Service) Started() bool {
+	return open, executing, nil
 }
 
-func (s *Service) SetBroadcaster(b Broadcaster) {
-	if s.broadcaster.Swap(b) != nil {
-		log.Info("Service broadcaster updated")
+func (s *ServiceClient) SetBroadcaster(b Broadcaster) {
+	s.command <- func(s *Service) {
+		old := s.broadcaster
+		s.broadcaster = b
+		if old != nil {
+			log.Info("Service broadcaster updated")
+		}
 	}
 }
 
-func (s *Service) SetChallengeVerifier(provider challenge_verifier.Verifier) {
+func (s *ServiceClient) SetChallengeVerifier(provider challenge_verifier.Verifier) {
 	s.challengeVerifier.Store(provider)
 }
 
-func (s *Service) executeRound(ctx context.Context, r *round) error {
-	s.executingRoundsMutex.Lock()
-	s.executingRounds[r.ID] = r
-	s.executingRoundsMutex.Unlock()
-
-	defer func() {
-		s.executingRoundsMutex.Lock()
-		delete(s.executingRounds, r.ID)
-		s.executingRoundsMutex.Unlock()
-	}()
-
+func executeRound(ctx context.Context, r *round, end time.Time, minMemoryLayer uint) roundResult {
+	logger := logging.FromContext(ctx).WithFields(log.String("round", r.ID))
+	logger.Info("executing until %v...", end)
 	start := time.Now()
-	end := s.genesis.
-		Add(s.cfg.EpochDuration * time.Duration(r.Epoch()+1)).
-		Add(s.cfg.PhaseShift).
-		Add(-s.cfg.CycleGap)
-
-	log.Info("Round %v executing until %v...", r.ID, end)
-
-	if err := r.execute(ctx, end, uint(s.minMemoryLayer)); err != nil {
-		return err
+	err := r.execute(ctx, end, uint(minMemoryLayer))
+	if err == nil {
+		logger.Info("execution ended, phi=%x, duration %v", r.execution.NIP.Root, time.Since(start))
+	} else {
+		logger.With().Warning("execution failed", log.Err(err))
 	}
 
-	log.Info("Round %v execution ended, phi=%x, duration %v", r.ID, r.execution.NIP.Root, time.Since(start))
-
-	return nil
+	return roundResult{round: r, err: err}
 }
 
-func (s *Service) Submit(ctx context.Context, challenge, signature []byte) (*round, []byte, error) {
-	logger := logging.FromContext(ctx)
-	if !s.Started() {
-		return nil, nil, ErrNotStarted
+func (s *ServiceClient) Submit(ctx context.Context, challenge, signature []byte) (string, []byte, error) {
+	if s == nil {
+		return "", nil, ErrNotStarted
 	}
+	logger := logging.FromContext(ctx)
 
 	logger.Debug("Received challenge")
 	// SAFETY: it will never panic as `s.ChallengeVerifier` is set in Start
 	verifier := s.challengeVerifier.Load().(challenge_verifier.Verifier)
 	result, err := verifier.Verify(ctx, challenge, signature)
 	if err != nil {
 		logger.With().Debug("challenge verification failed", log.Err(err))
-		return nil, nil, err
+		return "", nil, err
 	}
 	logger.With().Debug("verified challenge", log.String("hash", hex.EncodeToString(result.Hash)), log.String("node_id", hex.EncodeToString(result.NodeId)))
-	s.openRoundMutex.Lock()
-	r := s.openRound
-	err = r.submit(result.NodeId, result.Hash)
-	s.openRoundMutex.Unlock()
+
+	type response struct {
+		round string
+		err   error
+	}
+	done := make(chan response, 1)
+	s.command <- func(s *Service) {
```

Comment:
not familiar with the app structure in poet. is there a timeout mechanism here? it could block forever if the last command processing get stuck

Reply:
> If processing a previous command gets stuck, then it is a bug that needs fixing. I can, however, do a select on done and ctx.Done() to let the caller a way to "give up".

I would welcome that